tests/servers/rendering/test_shader_preprocessor.h

```diff
@@ -0,0 +1,254 @@
+// shader code.
+String compact_spaces(String &p_str) {
+	String compacted = p_str.strip_edges();
+	compacted = String(" ").join(compacted.split(" ", false));
+	compacted = String("\n").join(compacted.split("\n", false));
+	return compacted;
+}
+
+#define CHECK_SHADER_EQ(a, b) CHECK_EQ(compact_spaces(a), compact_spaces(b))
+
+TEST_CASE("[ShaderPreprocessor] Simple defines") {
+	String code(
+			"#define X 1.0 // comment\n"
+			"#define Y mix\n"
+			"#define Z X\n"
+			"\n"
+			"#define func0 \\\n"
+			"  vec3 my_fun(vec3 arg) {\\\n"
+			"    return pow(arg, 2.2);\\\n"
+			"  }\n"
+			"\n"
+			"func0\n"
+			"\n"
+			"fragment() {\n"
+			"  ALBEDO = vec3(X);\n"
+			"  float x = Y(0., Z, X);\n"
+			"  #undef X\n"
+			"  float X = x;\n"
+			"  x = -Z;\n"
+			"}\n");
+	String expected(
+			"vec3 my_fun(vec3 arg) { return pow(arg, 2.2); }\n"
+			"\n"
+			"fragment() {\n"
+			"  ALBEDO = vec3(1.0);\n"
+			"  float x = mix(0., 1.0, 1.0);\n"
+			"  float X = x;\n"
+			"  x = -X;\n"
+			"}\n");
+	String result;
+
+	ShaderPreprocessor preprocessor;
+	CHECK_EQ(preprocessor.preprocess(code, String("file.gdshader"), result), Error::OK);
+
+	CHECK_SHADER_EQ(result, expected);
+}
+
+TEST_CASE("[ShaderPreprocessor] Complex defines") {
+	String code(
+			"const float X = 2.0;\n"
+			"#define A(X) X*2.\n"
+			"#define X 1.0\n"
+			"#define Y Z(X, W)\n"
+			"#define Z max\n"
+			"#define C(X, Y) Z(A(Y), B(X))\n"
+			"#define W -X\n"
+			"#define B(X) X*3.\n"
+			"\n"
+			"fragment() {\n"
+			"  float x = Y;\n"
+			"  float y = C(5., 7.0);\n"
+			"}\n");
+	String expected(
+			"const float X = 2.0;\n"
+			"fragment() {\n"
+			"  float x = max(1.0, -1.0);\n"
+			"  float y = max(7.0*2., 5.*3.);\n"
+			"}\n");
+	String result;
+
+	ShaderPreprocessor preprocessor;
+	CHECK_EQ(preprocessor.preprocess(code, String("file.gdshader"), result), Error::OK);
+
+	CHECK_SHADER_EQ(result, expected);
+}
+
+TEST_CASE("[ShaderPreprocessor] Concatenation") {
+	String code(
+			"fragment() {\n"
+			"  #define X 1 // this is fine ##\n"
+			"  #define y 2\n"
+			"  #define z 3##.## 1## 4 ## 59\n"
+			"  #define Z(y) X ## y\n"
+			"  #define Z2(y) y##X\n"
+			"  #define W(y) X, y\n"
+			"  #define A(x) fl## oat a = 1##x ##.3  ##  x\n"
+			"  #define C(x, y) x##.##y\n"
+			"  float Z(y) = 1.2;\n"
+			"  float Z(z) = 2.3;\n"
+			"  float Z2(y) = z;\n"
+			"  float Z2(z) = 2.3;\n"
+			"  int b = max(W(3));\n"
+			"  A(9);\n"
+			"  Xy = C(X, y);\n"
+			"}\n");
+	String expected(
+			"fragment() {\n"
+			"  float Xy = 1.2;\n"
+			"  float Xz = 2.3;\n"
+			"  float yX = 3.1459;\n"
+			"  float zX = 2.3;\n"
+			"  int b = max(1, 3);\n"
+			"  float a = 19.39;\n"
+			"  Xy = 1.2;\n"
+			"}\n");
+	String result;
+
+	ShaderPreprocessor preprocessor;
+	CHECK_EQ(preprocessor.preprocess(code, String("file.gdshader"), result), Error::OK);
+
+	CHECK_SHADER_EQ(result, expected);
+}
+
+TEST_CASE_MAY_FAIL("[ShaderPreprocessor] Nested concatenation") {
```

Comment:
This TEST_CASE_MAY_FAIL() is bit ugly in the test output, printing a big red error and the failed comparison text in the console even though the full test suite still passes. It might be better to make it a normal TEST_CASE() and "fail succesfully" to keep the "noise" in the testing results output minimal. The comment there should be enough to clarify the test case.

Reply:
Before I discovered the TEST_CASE_MAY_FAIL macro I had it marked both as may_fail and skip, but I'll do as you suggest.